data/src/components/atom.rs

```diff
@@ -0,0 +1,423 @@
+use crate::mode::utils::not_found;
+use crate::serialisation::serialise;
+
+/// Representation of the [`Atom`] component in [`Prove`] mode
+#[perfect_derive(Clone, Debug)]
+struct ProveImpl<'normal, T> {
+    /// Previous value held by the component
+    ///
+    /// This is relevant because the proof generation mode will result in a Merkle proof that
+    /// contains the initial values. Hence, we need to keep track of the previous value.
+    previous: Source<'normal, T>,
+
+    /// Current value held by the component
+    ///
+    /// This is important because in proof generation mode, the value may be updated. Subsequent
+    /// reads should reflect the updated value.
+    current: Option<Box<T>>,
+
+    /// Whether the value has been read during proof generation
+    ///
+    /// This helps us decide whether the value should be included in the Merkle proof or not.
+    read: Cell<bool>,
+}
+
+/// Modal template for the [`Atom`] component
+///
+/// This type helps us pick the representation of [`Atom`] for each mode by implementing [`Modal`].
+struct AtomTemplate<T: ?Sized>(PhantomData<T>, Infallible);
+
+impl<T: 'static> Modal for AtomTemplate<T> {
+    type Normal = T;
+
+    type Prove<'normal> = ProveImpl<'normal, T>;
+
+    type Verify = Partial<T>;
+}
+
+/// Implementing modes support common operations on [`Atom`] components
+///
+/// The methods of the [`Atom`] type provide a more convenient interface to the functionality of
+/// this trait.
+pub trait AtomMode: Mode {
+    /// See [`Atom::new`].
+    fn new<T: 'static>(value: T) -> Atom<T, Self>;
+
+    /// Obtain an immutable reference to the current value held.
+    fn deref<T: 'static>(this: &Atom<T, Self>) -> &T;
+
+    /// Obtain a mutable reference to the current value held.
+    fn deref_mut<T: Clone + 'static>(this: &mut Atom<T, Self>) -> &mut T;
+
+    /// See [`Atom::write`].
+    fn write<T: 'static>(this: &mut Atom<T, Self>, value: T);
+}
+
+impl AtomMode for Normal {
+    fn new<T: 'static>(value: T) -> Atom<T, Self> {
+        Atom { atom: value }
+    }
+
+    fn deref<T: 'static>(this: &Atom<T, Self>) -> &T {
+        &this.atom
+    }
+
+    fn deref_mut<T: 'static>(this: &mut Atom<T, Self>) -> &mut T {
+        &mut this.atom
+    }
+
+    fn write<T: 'static>(this: &mut Atom<T, Self>, value: T) {
+        this.atom = value;
+    }
+}
+
+impl<'normal> AtomMode for Prove<'normal> {
+    fn new<T: 'static>(value: T) -> Atom<T, Self> {
+        Atom {
+            atom: ProveImpl {
+                previous: Source::Owned(Box::new(value)),
+                current: None,
+                read: Cell::new(false),
+            },
+        }
+    }
+
+    fn deref<T: 'static>(this: &Atom<T, Self>) -> &T {
+        // When proving, we need to ensure the value is marked as read such that it will be
+        // included in the proof. This ensures the verifier, which will also call `deref`, can see
+        // the value.
+        this.atom.read.set(true);
```

Comment:
I suppose my question would be - if we `write` to the atom before using `deref` or `deref_mut` - should it be marked as read? Presumably the previous value doesn't have to exist in the merkle tree at all, only the hash.

Though by the current look - we don't support this at the moment?

Reply:
Yes, we don't support this kind of optimisation yet. Reads or writes in any order are treated the same. It's definitely something we need to look into. The issue for this is RV-509.